File: BirthdayBot/BackgroundServices/BackgroundService.vb
''' <summary>
''' Base class for background services.
''' Background services are called periodically by another class.
''' </summary>
MustInherit Class BackgroundService
    Protected ReadOnly Property BotInstance As BirthdayBot

    Sub New(instance As BirthdayBot)
        BotInstance = instance
    End Sub

    Sub Log(message As String)
        Program.Log(Me.GetType().Name, message)
    End Sub

    MustOverride Function OnTick(tick As Integer) As Task
End Class

File: BackgroundServices/BackgroundService.cs
namespace BirthdayBot.BackgroundServices;
abstract class BackgroundService {
    protected static SemaphoreSlim ConcurrentSemaphore { get; private set; } = null!;

    protected ShardInstance Shard { get; }

    public BackgroundService(ShardInstance instance) {
        Shard = instance;
        ConcurrentSemaphore ??= new SemaphoreSlim(instance.Config.MaxConcurrentOperations);
    }

    protected void Log(string message) => Shard.Log(GetType().Name, message);

    public abstract Task OnTick(int tickCount, CancellationToken token);
}
